josephlatina/DungeonCrawler
# Room node graph architecture + Editor Window + Organizing Assets

## [x] **Organizing Assets**

- I moved all of the scriptable object assets into one folder `Scriptable Objects`
- I added Inventory System SO as a Game Resource

### **Adding a Game Resource and Accessing it**
- I added a Game Resource script and prefab instance which you can use to access any resources easily from any game component
- This is useful if you want to have a Scriptable Object Asset to exist as a single instance within a game session
- First, you modify `GameResources.cs`: Assets -> Scripts -> GameManager -> GameResources.cs
  - You will find a section for all of the resources variable. You just add a line declaring your resource variable: `public InventorySystem inventorySystem;`
- In the GameResources prefab instance I've created, you will see a field in the Inspector for the resource you declared. Simply drag your Scriptable Object to that field.
- To access the resource from any game component: `GameResources.Instance.x;` where x is your resource variable

### **Added a Settings Script**
- You can access it through: Assets -> Scripts -> Misc -> `Settings.cs`
- Purpose of this file is to store any configuration parameters we want to set for the game (ie. audio settings, control configuration)
- Currently I have dungeon room settings here that act as a constraint when I'm designing the node graphs

## [x] **Scriptable Objects/Scripts for every Node Graph Data Structure + Editor Window**
- I've added scripts for creating scriptable object assets for every data structure in the node graph hierarchy
- I also implemented an Editor Window GUI that can easily create scriptable objects for room node graphs and all of the room nodes within. This is useful for easily creating multiple variations of each level (graph) within the game.
- The Editor Window GUI also establishes the parent-child link between the room nodes as well as defining the room node type for each of the room nodes
- I've implemented modification functionalities (ie. deletion) within the GUI that enables user to easily modify the nodes and relationships within existing graphs (in case we want to modify them in the future)

### **_Testing the Node Graph Editor Window_**
- Go to Assets -> Scriptable Objects -> Dungeon Rooms -> DungeonRoomNodeGraphs
- Right click on the folder `DungeonRoomNodeGraphs` and create a scriptable object for Dungeon Room Node Graph
- Double click on the newly created RoomNodeGraph SO (it should open an editor window GUI)
- Right click on the canvas and `Create Room Node` (it should create the entrance node automatically and another room node)
- Click on the created room node's dropdown to set the room type for that node (try corridor first)
- Right click and drag from the Entrance node to the Corridor node (it should establish a parent-child link)
- Some of the constraints/validations I've implemented for creating graphs:
  - Room nodes have to have a corridor in between
  - There is only one boss room within the graph
  - Entrance node has to be the top parent level node in the graph
  - .... (other validations I've put in the script `RoomNodeSO.cs`)
- Other functionalities accessed by right clicking anywhere in the canvas:
  - Selecting all of the room nodes in the graph (you can also individually select nodes by manually left clicking on each node)
  - Deleting selected room nodes
  - Deleting the **links** between selected room nodes
- You can also click and drag the canvas to move around within the window
- **Note that for every room node you've added/setup within the Editor Window, you will see the assets folder get populated with scriptable objects for those room nodes. You will see them within the node graph SO you've initially created.**

<img width="803" alt="image" src="https://github.com/josephlatina/DungeonCrawler/assets/74115441/9088e63b-dba2-4691-9e55-ca6b824f909b">
